### ISSUE TITLE
Add dead store related warnings

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#24880]
|  |  |
| --- | --- |
| Bugzilla Link | [24506](https://llvm.org/bz24506) |
| Version | unspecified |
| OS | All |
| CC | @chandlerc,@DougGregor,@Weverything,@yurivict |

## Extended Description 
A while ago this bug came up:
```
void func(bool *b) {
  b = false;
}
```
Where the user forgot to dereference 'b' before assigning. This is now warned about (`-Wbool-conversion`) in < C++11, and a hard error in >= C++11 (C++11 and up, only the literal 0, not all integer zero literals, is a valid null pointer - so `'\0'` is also caught)

But this bug could still be written with integers:
```
void func1(int *i) {
  i = 0;
}
```
One way to catch this would be to implement a warning for dead stores. Any situation in which a variable is written to, but never read. (the zero was written, and then `i` went out of scope - thus never read)

This wouldn't catch another case that came up, though:
```
void func2(bool b) {
  b = false;
  f1(b);
  f2(b);
  ...
}
```
The author intended the boolean to be a reference and to communicate back to the caller some state about the execution of `f1`, `f2`, etc...

A dead store warning, extended somewhat, could catch this - if we treat the input parameter `b` as a store itself, then it is a dead store if we unconditionally write to `b` before we ever read from it. (as is the case `func2`)

It may be the case that a general dead store warning (even with the further generalization for parameter 'stores') would be too noisy for existing codebases, consider common idioms like:
```
f(x); ++x
f(x); ++x
```
etc. People copy/paste these, and they may intentionally leave the `++x` on the last line for consistency.

So to get a good warning we might need to blacklist certain store operations to be acceptably dead - or whitelist those (like parameter stores) that would never be acceptably dead.

GCC has a simpler form of such a warning which wouldn't require control flow analysis, `-Wunused-but-set-parameter`. Just looks for the parameter to be assigned to in the function, but never to be read from at all. This would catch the basic cases shown at the start, but not func2 or anything more involved. May be worth implementing if someone wants some coverage but doesn't want to pay the compile time penalty for CFG-based warnings.

**Comment by wheatman** [comment]
This first two issues seems to be solved between 12 and 13 with the new flag `-Wunused-but-set-parameter` which admittedly is not in `-Wall`
https://godbolt.org/z/GW1hYjdef

code
```
void func1(int *i) {
    i = 0;
}

void func(bool *b) {
b = false;
}

void func2(bool *b) {
b = 0;
}
```

new warnings

```
<source>:1:17: warning: parameter 'i' set but not used [-Wunused-but-set-parameter]
void func1(int *i) {
                ^
<source>:6:5: error: incompatible integer to pointer conversion assigning to 'bool *' from 'bool'
b = false;
    ^~~~~
<source>:9:18: warning: parameter 'b' set but not used [-Wunused-but-set-parameter]
void func2(bool *b) {
                 ^
2 warnings and 1 error generated.
Compiler returned: 1
```

For the last case, I am not very sure what you would want to happen
```
void func2(bool b) {
b = false;
f1(b);
f2(b);
}
```
is perfectively valid code, it might be a little weird to unconditionally set b to false, but what we even warn for

**Comment by dwblaikie** [comment]
> This first two issues seems to be solved between 12 and 13 with the new flag `-Wunused-but-set-parameter` which admittedly is not in `-Wall` https://godbolt.org/z/GW1hYjdef
> 
> code
> 
> ```
> void func1(int *i) {
>     i = 0;
> }
> 
> void func(bool *b) {
> b = false;
> }
> 
> void func2(bool *b) {
> b = 0;
> }
> ```
> 
> new warnings
> 
> ```
> <source>:1:17: warning: parameter 'i' set but not used [-Wunused-but-set-parameter]
> void func1(int *i) {
>                 ^
> <source>:6:5: error: incompatible integer to pointer conversion assigning to 'bool *' from 'bool'
> b = false;
>     ^~~~~
> <source>:9:18: warning: parameter 'b' set but not used [-Wunused-but-set-parameter]
> void func2(bool *b) {
>                  ^
> 2 warnings and 1 error generated.
> Compiler returned: 1
> ```

I'd still like to see this bug kept open for the more general not-just-parameters cases like this. 

> For the last case, I am not very sure what you would want to happen
> 
> ```
> void func2(bool b) {
> b = false;
> f1(b);
> f2(b);
> }
> ```
> 
> is perfectively valid code, it might be a little weird to unconditionally set b to false, but what we even warn for

Most of the code we warn on is perfectly valid - but suspicious/probably not what someone intended.

If this is what the user intends, likely they would've written it as:
```
void func2(bool) { // assuming they needed the signature for some reason, omitting the name is generally how we communicate that it's intended to be unused
  bool b = false;
  f1(b);
  f2(b);
}
```
